text/0244-eks-developer-preview-migration-path.md

```diff
@@ -0,0 +1,102 @@
+strategy.
+
+# Design Summary
+
+The main idea is to spin off to a new module that introduces all the breaking changes at once. The old module will be effectively
+locked, and won't receive any updates apart from **p0** bug fixes. Its essentially the same as rolling out a new major version,
+but since all modules are versioned together, we do this by changing the module name, and not the version.
+
+Since the final module we mark as `stable` should be called `aws-eks`, we rename the current module to `aws-eks-experimental`,
+and move forward with the new one as `aws-eks`.
+
+This means that upon upgrade, existing users of `aws-eks` will now receive the developer preview version, which will naively replace their clusters.
+To avoid that, we need the code to detect this situation and provide clear instructions to the user.
+
+# README
+
+In the old module, which is now `aws-eks-experimental`, we will add the following notice to the README.
+
+> This module has been deprecated in favor of the `aws-eks` module. This module will not receive updates apart from critical bug fixes.
+> Note that transitioning to `aws-eks` might incur changes that require cluster replacement.
+
+In the new module, which is now `aws-eks`, we will add the following notice to the README.
+
+> If you are transitioning to this module from `aws-eks-experimental`, you might incur changes that require cluster replacements.
+
+# Implementation Plan
+
+1. Rename current module to `aws-eks-experimental` on the master branch.
+
+    - This will be released with a BREAKING CHANGE notice and cause users to change their dependency.
+
+2. Create a staging branch `eks-dev-preview` which contains `aws-eks` as a copy of `aws-eks-experimental`.
+
+    - This will be a long living branch we will branch out of to and merge to.
+
+3. Implement necessary changes onto the `eks-dev-preview` branch.
+
+    - We branch out of the `aws-dev-preview` branch to separate feature branches so that we can review the changes.
+
+4. Merge `eks-dev-preview` to master.
+
+    - This merge is the release of the developer preview module.
+
+# Rationale and Alternatives
+
+## Rename `eks.Cluster`
+
+Instead of spinning off to a new module, we can spin off to a new construct.
+
+That is, rename the current `eks.Cluster` to `eks.ExperimentalCluster` and implement everything we need in `eks.Cluster`.
+We have actually already done this once with `eks.LegacyCluster`.
+
+The benefit here is that users won't need to change dependencies. However, this change would also require a name change and
+duplication of any class that interacts with `eks.Cluster` since those classes may not work with both `eks.Cluster` and `eks.ExperimentalCluster`.
+
+This approach feels more fragile and doesn't provide much value since the user would still incur a breaking change.
+It would just be in the form of a class name change, instead of a module name change.
+
+## Rename new module
```

Comment:
Alternately, we could put the new developer preview module in a separate repository and vend it as an independent package. When we go to GA with CDK 2, the old module could be replaced entirely. The CLI warning is probably still a good idea. This way, anyone who does not want to replace their clusters stays on 1.x until they are ready.

Reply:
What is the benefit of putting it in a different repo as opposed to just using a different name in the current repo? Also, are you suggesting the EKS library will be vended separately from the CDK until we release V2? Under which name? It feels wrong from a user's perspective.